stratum/hal/lib/barefoot/bfrt_table_manager.cc

```diff
@@ -678,6 +704,195 @@ ::util::Status BfrtTableManager::ReadTableEntry(
+
+  if (type == ::p4::v1::Update::DELETE) {
+    // Fetch and delete the table entry pointing to the action profile group ID.
+    ASSIGN_OR_RETURN(auto table_key,
+                     bf_sde_interface_->CreateTableKey(table_id));
+    RETURN_IF_ERROR(BuildTableKey(table_entry, table_key.get()));
+    ASSIGN_OR_RETURN(auto table_data,
+                     bf_sde_interface_->CreateTableData(
+                         table_id, table_entry.action().action().action_id()));
+    RETURN_IF_ERROR(bf_sde_interface_->GetTableEntry(
+        device_, session, table_id, table_key.get(), table_data.get()));
+    uint64 action_profile_group_id;
+    RETURN_IF_ERROR(table_data->GetSelectorGroupId(&action_profile_group_id));
+    RETURN_IF_ERROR(bf_sde_interface_->DeleteTableEntry(
+        device_, session, table_id, table_key.get()));
+    // Fetch and delete the action profile group containing the members.
+    std::vector<int> group_ids;
+    std::vector<int> max_group_sizes;
+    std::vector<std::vector<uint32>> member_ids;
+    std::vector<std::vector<bool>> member_statuses;
+    RETURN_IF_ERROR(bf_sde_interface_->GetActionProfileGroups(
+        device_, session, bfrt_act_sel_table_id, action_profile_group_id,
+        &group_ids, &max_group_sizes, &member_ids, &member_statuses));
+    CHECK_EQ(1, group_ids.size());
+    CHECK_EQ(1, member_ids.size());
+    CHECK_EQ(action_profile_group_id, group_ids[0]);
+    RETURN_IF_ERROR(bf_sde_interface_->DeleteActionProfileGroup(
+        device_, session, bfrt_act_sel_table_id, action_profile_group_id));
+    // Delete the members.
+    for (const auto& member_id : member_ids[0]) {
+      RETURN_IF_ERROR(bf_sde_interface_->DeleteActionProfileMember(
+          device_, session, bfrt_act_prof_table_id, member_id));
+    }
+  } else if (type == ::p4::v1::Update::INSERT) {
+    // 1. Find free member and group IDs.
+    // TODO(max): improve ID search from linear extension to hole finding.
+    uint32 next_free_member_id = 1;
+    uint32 next_free_group_id = 1;
+    {
+      std::vector<int> all_group_ids;
+      std::vector<int> max_group_sizes;
+      std::vector<std::vector<uint32>> member_ids;
+      std::vector<std::vector<bool>> member_statuses;
+      RETURN_IF_ERROR(bf_sde_interface_->GetActionProfileGroups(
+          device_, session, bfrt_act_sel_table_id, 0, &all_group_ids,
+          &max_group_sizes, &member_ids, &member_statuses));
+      std::sort(all_group_ids.begin(), all_group_ids.end());
+      if (!all_group_ids.empty()) {
+        next_free_group_id = *(all_group_ids.rbegin()) + 1;
+      }
+      std::vector<uint32> all_member_ids;
+      for (const auto& member_group : member_ids) {
+        all_member_ids.insert(all_member_ids.end(), member_group.begin(),
+                              member_group.end());
+      }
+      std::sort(all_member_ids.begin(), all_member_ids.end());
+      if (!all_member_ids.empty()) {
+        next_free_member_id = *(all_member_ids.rbegin()) + 1;
+      }
```

Comment:
This search will eventually overflow back to 0 and potentially fail with a duplicate.